// react-jinke-music-player/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.CircleProcessBar = exports.PlayModel = exports.Load = exports.AnimatePauseIcon = exports.AnimatePlayIcon = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _isMobile = _interopRequireDefault(require("is-mobile"));

var _Slider = _interopRequireDefault(require("rc-slider/lib/Slider"));

var _rcSwitch = _interopRequireDefault(require("rc-switch"));

var _utils = require("./utils");

var _AudioListsPanel = _interopRequireDefault(require("./components/AudioListsPanel"));

var _PlayerMobile = _interopRequireDefault(require("./components/PlayerMobile"));

var _reactDraggable = _interopRequireDefault(require("react-draggable"));

var _headphones = _interopRequireDefault(require("react-icons/lib/fa/headphones"));

var _minusSquareO = _interopRequireDefault(require("react-icons/lib/fa/minus-square-o"));

var _playCircle = _interopRequireDefault(require("react-icons/lib/fa/play-circle"));

var _pauseCircle = _interopRequireDefault(require("react-icons/lib/fa/pause-circle"));

var _refresh = _interopRequireDefault(require("react-icons/lib/fa/refresh"));

var _volumeUp = _interopRequireDefault(require("react-icons/lib/md/volume-up"));

var _volumeOff = _interopRequireDefault(require("react-icons/lib/md/volume-off"));

var _cloudDownload = _interopRequireDefault(require("react-icons/lib/fa/cloud-download"));

var _spinner = _interopRequireDefault(require("react-icons/lib/fa/spinner"));

var _repeatOne = _interopRequireDefault(require("react-icons/lib/md/repeat-one"));

var _repeat = _interopRequireDefault(require("react-icons/lib/md/repeat"));

var _arrowShuffle = _interopRequireDefault(require("react-icons/lib/ti/arrow-shuffle"));

var _viewHeadline = _interopRequireDefault(require("react-icons/lib/md/view-headline"));

var _queueMusic = _interopRequireDefault(require("react-icons/lib/md/queue-music"));

var _skipNext = _interopRequireDefault(require("react-icons/lib/md/skip-next"));

var _skipPrevious = _interopRequireDefault(require("react-icons/lib/md/skip-previous"));

var _close = _interopRequireDefault(require("react-icons/lib/md/close"));

var _trashO = _interopRequireDefault(require("react-icons/lib/fa/trash-o"));

require("rc-slider/assets/index.css");

require("rc-switch/assets/index.css");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

var ISMOBILE = (0, _isMobile.default)();

var AnimatePlayIcon = function AnimatePlayIcon() {
  return _react.default.createElement(_playCircle.default, {
    className: "react-jinke-music-player-play-icon"
  });
};

exports.AnimatePlayIcon = AnimatePlayIcon;

var AnimatePauseIcon = function AnimatePauseIcon() {
  return _react.default.createElement(_pauseCircle.default, {
    className: "react-jinke-music-player-pause-icon"
  });
};

exports.AnimatePauseIcon = AnimatePauseIcon;

var Load = function Load() {
  return _react.default.createElement("span", {
    className: "loading group",
    key: "loading"
  }, _react.default.createElement(_spinner.default, null));
};

exports.Load = Load;

var PlayModel = function PlayModel(_ref) {
  var visible = _ref.visible,
      value = _ref.value;
  return _react.default.createElement("div", {
    className: (0, _classnames.default)("play-mode-title", {
      "play-mode-title-visible": visible
    }),
    key: "play-mode-title"
  }, value);
}; //迷你模式进度条


exports.PlayModel = PlayModel;

var CircleProcessBar = function CircleProcessBar() {
  var _ref2 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
      _ref2$progress = _ref2.progress,
      progress = _ref2$progress === void 0 ? 0 : _ref2$progress,
      _ref2$r = _ref2.r,
      r = _ref2$r === void 0 ? 40 : _ref2$r;

  var _progress = progress.toFixed(2);

  var perimeter = Math.PI * 2 * r;
  var strokeDasharray = "".concat(~~(perimeter * _progress), " ").concat(~~(perimeter * (1 - _progress)));
  return _react.default.createElement("svg", {
    className: "audio-circle-process-bar"
  }, _react.default.createElement("circle", {
    cx: r,
    cy: r,
    r: r - 1,
    fill: "none",
    className: "stroke",
    strokeDasharray: strokeDasharray
  }), _react.default.createElement("circle", {
    cx: r,
    cy: r,
    r: r - 1,
    fill: "none",
    className: "bg",
    strokeDasharray: "0 1000"
  }));
};

exports.CircleProcessBar = CircleProcessBar;
var sliderBaseOptions = {
  min: 0,
  step: 0.01,
  trackStyle: {
    backgroundColor: "#31c27c"
  },
  handleStyle: {
    backgroundColor: "#31c27c",
    border: "2px solid #fff"
  }
};

var ReactJkMusicPlayer =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(ReactJkMusicPlayer, _PureComponent);

  //初始播放id
  // 列表删除动画时间(ms)
  function ReactJkMusicPlayer(props) {
    var _this;

    _classCallCheck(this, ReactJkMusicPlayer);

    _this = _possibleConstructorReturn(this, (ReactJkMusicPlayer.__proto__ || Object.getPrototypeOf(ReactJkMusicPlayer)).call(this, props));
    Object.defineProperty(_assertThisInitialized(_this), "initPlayId", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: -1
    });
    Object.defineProperty(_assertThisInitialized(_this), "audioListRemoveAnimateTime", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: 350
    });
    Object.defineProperty(_assertThisInitialized(_this), "NETWORK_STATE", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: {
        NETWORK_EMPTY: 0,
        //未初始化
        NETWORK_IDLE: 1,
        //未使用网络 304 缓存
        NETWORK_LOADING: 2,
        //浏览器正在下载数据
        NETWORK_NO_SOURCE: 3 //未找到资源

      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "READY_SUCCESS_STATE", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: 4
    });
    Object.defineProperty(_assertThisInitialized(_this), "state", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: {
        audioLists: [],
        playId: _this.initPlayId,
        //播放id
        name: "",
        //当前歌曲名
        cover: "",
        //当前歌曲封面
        singer: "",
        //当前歌手
        musicSrc: "",
        //当前歌曲链接
        isMobile: ISMOBILE,
        toggle: false,
        pause: true,
        playing: false,
        duration: 0,
        currentTime: 0,
        isLoop: false,
        isMute: false,
        soundValue: 100,
        isDrag: false,
        //是否支持拖拽
        currentX: 0,
        currentY: 0,
        moveX: 0,
        moveY: 0,
        isMove: false,
        loading: false,
        audioListsPanelVisible: false,
        playModelNameVisible: false,
        theme: _this.darkThemeName,
        extendsContent: [],
        //自定义扩展功能按钮
        playMode: "",
        //当前播放模式
        currentAudioVolume: 0,
        //当前音量  静音后恢复到之前记录的音量
        initAnimate: false,
        isInitAutoplay: false,
        isInitRemember: false,
        loadProgress: 0,
        removeId: -1
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "togglePlayMode", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        var index = _this._PLAY_MODE_.findIndex(function (_ref3) {
          var key = _ref3.key;
          return key === _this.state.playMode;
        });

        var playMode = index === _this._PLAY_MODE_LENGTH_ - 1 ? _this._PLAY_MODE_[0]["key"] : _this._PLAY_MODE_[++index]["key"];

        _this.setState({
          playMode: playMode,
          playModelNameVisible: true,
          playModeTipVisible: true
        });

        _this.props.onPlayModeChange && _this.props.onPlayModeChange(_this.PLAYMODE[playMode]);
        clearTimeout(_this.playModelTimer);
        _this.playModelTimer = setTimeout(function () {
          _this.setState({
            playModelNameVisible: false,
            playModeTipVisible: false
          });
        }, _this.props.playModeShowTime);
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "renderPlayModeIcon", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value(playMode) {
        var IconNode = "";
        var animateName = "react-jinke-music-player-mode-icon";

        switch (playMode) {
          case _this.PLAYMODE["order"]["key"]:
            IconNode = _react.default.createElement(_viewHeadline.default, {
              className: animateName
            });
            break;

          case _this.PLAYMODE["orderLoop"]["key"]:
            IconNode = _react.default.createElement(_repeat.default, {
              className: animateName
            });
            break;

          case _this.PLAYMODE["singleLoop"]["key"]:
            IconNode = _react.default.createElement(_repeatOne.default, {
              className: animateName
            });
            break;

          case _this.PLAYMODE["shufflePlay"]["key"]:
            IconNode = _react.default.createElement(_arrowShuffle.default, {
              className: animateName
            });
            break;

          default:
            IconNode = _react.default.createElement(_viewHeadline.default, {
              className: animateName
            });
        }

        return IconNode;
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "audioListsPlay", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value(playId) {
        var ignore = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
        var _this$state = _this.state,
            currentPlayId = _this$state.playId,
            pause = _this$state.pause,
            playing = _this$state.playing,
            audioLists = _this$state.audioLists;

        if (Array.isArray(audioLists) && audioLists.length === 0) {
          /*eslint-disable no-console*/
          return console.warn("Your playlist has no songs. and cannot play !");
          /*eslint-disable no-console*/
        } //如果点击当前项 就暂停 或者播放


        if (playId === currentPlayId && !ignore) {
          _this.setState({
            pause: !pause,
            playing: !playing
          });

          return pause ? _this.audio.play() : _this._pauseAudio();
        }

        var _audioLists$playId = audioLists[playId],
            name = _audioLists$playId.name,
            cover = _audioLists$playId.cover,
            musicSrc = _audioLists$playId.musicSrc,
            singer = _audioLists$playId.singer;

        _this.setState({
          name: name,
          cover: cover,
          musicSrc: musicSrc,
          singer: singer,
          playId: playId,
          currentTime: 0,
          duration: 0,
          playing: false,
          loading: true,
          loadProgress: 0
        }, function () {
          _this.audio.load();
        });

        _this.props.onAudioPlayTrackChange && _this.props.onAudioPlayTrackChange(playId, audioLists, _this.getBaseAudioInfo());
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "clearAudioDuration", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        _this.setState({
          currentTime: 0,
          duration: 0,
          loading: false,
          playing: false,
          pause: true
        });
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "deleteAudioLists", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value(id) {
        return function (e) {
          e.stopPropagation(); //如果不 传 id  删除全部

          var _this$state2 = _this.state,
              audioLists = _this$state2.audioLists,
              playId = _this$state2.playId;

          if (audioLists.length < 1) {
            return;
          }

          var newAudioLists = _toConsumableArray(audioLists).filter(function (_, i) {
            return i !== id;
          });

          if (!id && id !== 0) {
            _this.initPlayInfo([]);

            _this.clearAudioDuration();
          } else {
            //触发删除动画,等动画结束 删除列表
            _this.setState({
              removeId: id
            });

            setTimeout(function () {
              _this.setState({
                audioLists: newAudioLists,
                removeId: -1
              }, function () {
                if (id === playId) {
                  _this.initPlayInfo(newAudioLists);

                  _this.onPlay();

                  _this.props.onAudioPlayTrackChange && _this.props.onAudioPlayTrackChange(playId, newAudioLists, _this.getBaseAudioInfo());
                }

                if (newAudioLists.length <= 0) {
                  _this.clearAudioDuration();
                }
              });
            }, _this.audioListRemoveAnimateTime);
          }

          _this.props.onAudioListsChange && _this.props.onAudioListsChange(playId, newAudioLists, _this.getBaseAudioInfo());
        };
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "openAudioListsPanel", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        _this.setState(function (_ref4) {
          var audioListsPanelVisible = _ref4.audioListsPanelVisible,
              initAnimate = _ref4.initAnimate;
          return {
            initAnimate: true,
            audioListsPanelVisible: !audioListsPanelVisible
          };
        });

        _this.props.onAudioListsPanelChange && _this.props.onAudioListsPanelChange(!_this.state.audioListsPanelVisible);
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "closeAudioListsPanel", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value(e) {
        e.stopPropagation();

        _this.setState({
          audioListsPanelVisible: false
        });

        _this.props.onAudioListsPanelChange && _this.props.onAudioListsPanelChange(false);
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "themeChange", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value(isLight) {
        var theme = isLight ? _this.lightThemeName : _this.darkThemeName;

        _this.setState({
          theme: theme
        });

        _this.props.onThemeChange && _this.props.onThemeChange(theme);
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "onAudioDownload", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        var _this$state3 = _this.state,
            name = _this$state3.name,
            musicSrc = _this$state3.musicSrc;

        if (name && musicSrc) {
          _this.downloadNode = document.createElement("a");

          _this.downloadNode.setAttribute("download", name);

          _this.downloadNode.setAttribute("name", name);

          _this.downloadNode.setAttribute("href", musicSrc);

          _this.downloadNode.click();

          _this.downloadNode = undefined;
          _this.props.onAudioDownload && _this.props.onAudioDownload(_this.getBaseAudioInfo());
        }
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "controllerMouseMove", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value(e, _ref5) {
        var deltaX = _ref5.deltaX,
            deltaY = _ref5.deltaY;

        var isMove = Math.abs(deltaX) >= _this.openPanelPeriphery || Math.abs(deltaY) >= _this.openPanelPeriphery;

        _this.setState({
          isMove: isMove
        });
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "controllerMouseUp", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value(e, _ref6) {
        var x = _ref6.x,
            y = _ref6.y;

        if (!_this.state.isMove) {
          _this.openPanel();
        }

        _this.setState({
          moveX: x,
          moveY: y
        });

        return false;
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "controllerMouseOut", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value(e) {
        e.preventDefault();
        _this.isDrag = false;
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "onHandleProgress", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value(_value) {
        _this.audio.currentTime = _value;
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "onSound", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        _this.setAudioVolume(_this.state.currentAudioVolume);
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "setAudioVolume", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value(_value2) {
        _this.audio.volume = _value2;

        _this.setState({
          currentAudioVolume: _value2,
          soundValue: _value2
        });
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "stopAll", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value(target) {
        target.stopPropagation();
        target.preventDefault();
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "getBoundingClientRect", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value(ele) {
        var _ele$getBoundingClien = ele.getBoundingClientRect(),
            left = _ele$getBoundingClien.left,
            top = _ele$getBoundingClien.top;

        return {
          left: left,
          top: top
        };
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "audioLoop", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        _this.setState(function (_ref7) {
          var isLoop = _ref7.isLoop;
          return {
            isLoop: !isLoop
          };
        });
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "onAudioReload", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        _this.handlePlay(_this.PLAYMODE["singleLoop"]["key"]);

        _this.props.onAudioReload && _this.props.onAudioReload(_this.getBaseAudioInfo());
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "openPanel", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        if (_this.props.toggleMode) {
          _this.setState({
            toggle: true
          });

          _this.props.onModeChange && _this.props.onModeChange("full");
        }
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "onHidePanel", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value(e) {
        _this.setState({
          toggle: false,
          audioListsPanelVisible: false
        });

        _this.props.onModeChange && _this.props.onModeChange("mini");
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "onPlay", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        if (_this.state.audioLists.length >= 1) {
          //是否在播放
          var playing = _this.state.playing;

          if (playing === true) {
            _this._pauseAudio();
          } else {
            _this.setAudioLength();

            _this.loadAndPlayAudio();
          }
        }
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "_pauseAudio", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        _this.audio.pause();

        _this.setState({
          playing: false,
          pause: true
        });
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "onPauseAudio", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        _this.props.onAudioPause && _this.props.onAudioPause(_this.getBaseAudioInfo());
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "loadAndPlayAudio", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        var _this$props = _this.props,
            autoPlay = _this$props.autoPlay,
            remember = _this$props.remember;
        var _this$state4 = _this.state,
            isInitAutoplay = _this$state4.isInitAutoplay,
            isInitRemember = _this$state4.isInitRemember,
            loadProgress = _this$state4.loadProgress;
        var _this$audio = _this.audio,
            readyState = _this$audio.readyState,
            networkState = _this$audio.networkState;
        var maxLoadProgress = 100;

        _this.setState({
          loading: true
        });

        if (loadProgress < maxLoadProgress) {
          _this.setState({
            loadProgress: loadProgress + 1
          });
        }

        if (readyState === _this.READY_SUCCESS_STATE && networkState !== _this.NETWORK_STATE.NETWORK_NO_SOURCE) {
          var _this$getLastPlayStat = _this.getLastPlayStatus(),
              pause = _this$getLastPlayStat.pause;

          var isLastPause = remember && !isInitRemember && pause;
          var canPlay = isInitAutoplay || autoPlay === true;

          _this.setState({
            playing: remember ? !isLastPause : canPlay,
            loading: false,
            pause: remember ? isLastPause : !canPlay,
            loadProgress: maxLoadProgress
          }, function () {
            if (remember ? !isLastPause : canPlay) {
              _this.audio.play();
            }

            _this.setState({
              isInitAutoplay: true,
              isInitRemember: true
            });
          });
        }
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "setAudioLength", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        _this.setState({
          duration: _this.audio.duration
        });
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "onAudioLoadError", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value(e) {
        var _this$state5 = _this.state,
            playMode = _this$state5.playMode,
            audioLists = _this$state5.audioLists,
            playId = _this$state5.playId;

        if (audioLists.length >= 1) {
          //如果当前音乐加载出错 尝试播放下一首
          var loadAudioErrorPlayNext = _this.props.loadAudioErrorPlayNext;

          if (loadAudioErrorPlayNext && playId < audioLists.length - 1) {
            _this.handlePlay(playMode);
          }

          var info = _this.getBaseAudioInfo();

          _this.props.onAudioLoadError && _this.props.onAudioLoadError(_objectSpread({}, e, {
            audioInfo: info,
            errMsg: _this.audio.error || null
          }));
        }
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "handlePlay", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value(playMode) {
        var isNext = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
        var IconNode = "";
        var playId = _this.state.playId;
        var audioListsLen = _this.state.audioLists.length;

        switch (playMode) {
          //顺序播放
          case _this.PLAYMODE["order"]["key"]:
            IconNode = _react.default.createElement(_viewHeadline.default, null);
            if (playId === audioListsLen - 1) return _this._pauseAudio();

            _this.audioListsPlay(isNext ? ++playId : --playId);

            break;
          //列表循环

          case _this.PLAYMODE["orderLoop"]["key"]:
            IconNode = _react.default.createElement(_repeat.default, null);

            if (isNext) {
              if (playId === audioListsLen - 1) playId = _this.initPlayId;

              _this.audioListsPlay(++playId);
            } else {
              if (playId - 1 === _this.initPlayId) playId = audioListsLen;

              _this.audioListsPlay(--playId);
            }

            break;
          //单曲循环

          case _this.PLAYMODE["singleLoop"]["key"]:
            IconNode = _react.default.createElement(_repeatOne.default, null);
            _this.audio.currentTime = 0;

            _this.audioListsPlay(playId, true);

            break;
          //随机播放

          case _this.PLAYMODE["shufflePlay"]["key"]:
            {
              IconNode = _react.default.createElement(_arrowShuffle.default, null);
              var randomPlayId = (0, _utils.createRandomNum)(0, audioListsLen - 1);

              _this.audioListsPlay(randomPlayId, true);
            }
            break;

          default:
            IconNode = _react.default.createElement(_viewHeadline.default, null);
        }
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "audioEnd", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        _this.props.onAudioEnded && _this.props.onAudioEnded(_this.getBaseAudioInfo());

        _this.handlePlay(_this.state.playMode);
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "audioPrevAndNextBasePlayHandle", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        var isNext = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : true;
        var playMode = _this.state.playMode;
        var _playMode = "";

        switch (playMode) {
          case _this.PLAYMODE["shufflePlay"]["key"]:
            _playMode = playMode;
            break;

          default:
            _playMode = _this.PLAYMODE["orderLoop"]["key"];
            break;
        }

        _this.handlePlay(_playMode, isNext);
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "audioPrevPlay", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        _this.audioPrevAndNextBasePlayHandle(false);
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "audioNextPlay", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        _this.audioPrevAndNextBasePlayHandle(true);
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "audioTimeUpdate", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        var currentTime = _this.audio.currentTime;

        _this.setState({
          currentTime: currentTime
        });

        if (_this.props.remember) {
          _this.saveLastPlayStatus();
        }

        _this.props.onAudioProgress && _this.props.onAudioProgress(_this.getBaseAudioInfo());
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "audioSoundChange", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value(_value3) {
        _this.setAudioVolume(_value3);
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "onAudioVolumeChange", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        _this.setState({
          isMute: _this.audio.volume <= 0
        });

        _this.props.onAudioVolumeChange && _this.props.onAudioVolumeChange(_this.audio.volume);
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "onAudioPlay", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        _this.props.onAudioPlay && _this.props.onAudioPlay(_this.getBaseAudioInfo());
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "onAudioSeeked", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        if (_this.state.audioLists.length >= 1) {
          _this.loadAndPlayAudio();

          _this.props.onAudioSeeked && _this.props.onAudioSeeked(_this.getBaseAudioInfo());
        }
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "onMute", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        _this.setState({
          isMute: true,
          soundValue: 0,
          currentAudioVolume: _this.audio.volume
        }, function () {
          _this.audio.volume = 0;
        });
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "onAudioAbort", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value(e) {
        var audioInfo = _this.getBaseAudioInfo();

        var _err = Object.assign({}, e, audioInfo);

        _this.props.onAudioAbort && _this.props.onAudioAbort(_err);
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "toggleMode", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value(mode) {
        if (mode === _this.toggleModeName["full"]) {
          _this.setState({
            toggle: true
          });
        }
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "audioListsDragEnd", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value(fromIndex, toIndex) {
        var _this$state6 = _this.state,
            playId = _this$state6.playId,
            audioLists = _this$state6.audioLists;

        var _audioLists = _toConsumableArray(audioLists);

        var item = _audioLists.splice(fromIndex, 1)[0];

        _audioLists.splice(toIndex, 0, item); //如果拖动正在播放的歌曲 播放Id 等于 拖动后的index


        var _playId = fromIndex === playId ? toIndex : playId;

        _this.setState({
          audioLists: _audioLists,
          playId: _playId
        });

        _this.props.onAudioListsDragEnd && _this.props.onAudioListsDragEnd(fromIndex, toIndex);
        _this.props.onAudioListsChange && _this.props.onAudioListsChange(_playId, _audioLists, _this.getBaseAudioInfo());
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "saveLastPlayStatus", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        var _this$state7 = _this.state,
            currentTime = _this$state7.currentTime,
            playId = _this$state7.playId,
            duration = _this$state7.duration,
            theme = _this$state7.theme,
            soundValue = _this$state7.soundValue,
            playMode = _this$state7.playMode,
            name = _this$state7.name,
            cover = _this$state7.cover,
            singer = _this$state7.singer,
            musicSrc = _this$state7.musicSrc,
            pause = _this$state7.pause;
        var lastPlayStatus = JSON.stringify({
          currentTime: currentTime,
          playId: playId,
          duration: duration,
          theme: theme,
          playMode: playMode,
          soundValue: soundValue,
          name: name,
          cover: cover,
          singer: singer,
          musicSrc: musicSrc,
          pause: pause
        });
        localStorage.setItem("lastPlayStatus", lastPlayStatus);
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "getLastPlayStatus", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        var _this$props2 = _this.props,
            theme = _this$props2.theme,
            defaultPlayMode = _this$props2.defaultPlayMode,
            defaultPlayIndex = _this$props2.defaultPlayIndex;
        var status = {
          currentTime: 0,
          duration: 0,
          playMode: defaultPlayMode,
          name: "",
          cover: "",
          singer: "",
          musicSrc: "",
          playId: _this.getDefaultPlayIndex(defaultPlayIndex),
          theme: theme,
          pause: false
        };

        try {
          return JSON.parse(localStorage.getItem("lastPlayStatus")) || status;
        } catch (error) {
          return status;
        }
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "bindMobileAutoPlayerEvents", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        document.addEventListener("touchstart", function () {
          _this.audio.load();

          _this.onPlay();
        }, {
          once: true
        }); //监听微信准备就绪事件

        document.addEventListener("WeixinJSBridgeReady", function () {
          _this.audio.load();

          _this.onPlay();
        });
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "unBindEvents", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        var _this2;

        (_this2 = _this).bindEvents.apply(_this2, arguments);
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "bindEvents", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        var target = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : _this.audio;
        var eventsNames = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {
          waiting: _this.loadAndPlayAudio,
          canplay: _this.onPlay,
          error: _this.onAudioLoadError,
          ended: _this.audioEnd,
          seeked: _this.onAudioSeeked,
          pause: _this.onPauseAudio,
          play: _this.onAudioPlay,
          timeupdate: _this.audioTimeUpdate,
          volumechange: _this.onAudioVolumeChange,
          stalled: _this.onAudioLoadError,
          //当浏览器尝试获取媒体数据，但数据不可用时
          abort: _this.onAudioAbort
        };
        var bind = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
        var once = _this.props.once;

        for (var name in eventsNames) {
          var _events = eventsNames[name];
          bind ? target.addEventListener(name, _events, {
            once: !!(once && name === "play")
          }) : target.removeEventListener(name, _events);
        }
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "getPlayInfo", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        var audioLists = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
        var playId = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : _this.state.playId;

        var _ref8 = audioLists[Math.max(0, playId)] || {},
            _ref8$name = _ref8.name,
            name = _ref8$name === void 0 ? "" : _ref8$name,
            _ref8$cover = _ref8.cover,
            cover = _ref8$cover === void 0 ? "" : _ref8$cover,
            _ref8$singer = _ref8.singer,
            singer = _ref8$singer === void 0 ? "" : _ref8$singer,
            _ref8$musicSrc = _ref8.musicSrc,
            musicSrc = _ref8$musicSrc === void 0 ? "" : _ref8$musicSrc;

        return {
          name: name,
          cover: cover,
          singer: singer,
          musicSrc: musicSrc,
          audioLists: audioLists
        };
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "initPlayInfo", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value(audioLists) {
        var playId = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : _this.state.playId;
        var cb = arguments.length > 2 ? arguments[2] : undefined;

        _this.setState(_this.getPlayInfo(audioLists, playId), cb);
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "listenerIsMobile", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value(_ref9) {
        var matches = _ref9.matches;

        _this.setState({
          isMobile: !!matches
        });
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "addMobileListener", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        _this.media = window.matchMedia("(max-width: 768px) and (orientation : portrait)");

        _this.media.addListener(_this.listenerIsMobile);
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "setDefaultAudioVolume", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        var _this$props3 = _this.props,
            defaultVolume = _this$props3.defaultVolume,
            remember = _this$props3.remember; //音量 [0-100]

        _this.defaultVolume = Math.max(0, Math.min(defaultVolume, 100)) / 100;

        var _this$getLastPlayStat2 = _this.getLastPlayStatus(),
            _this$getLastPlayStat3 = _this$getLastPlayStat2.soundValue,
            soundValue = _this$getLastPlayStat3 === void 0 ? _this.defaultVolume : _this$getLastPlayStat3;

        _this.setAudioVolume(remember ? soundValue : _this.defaultVolume);
      }
    });
    Object.defineProperty(_assertThisInitialized(_this), "getDefaultPlayIndex", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: function value() {
        return Math.max(0, Math.min(_this.props.audioLists.length, _this.props.defaultPlayIndex));
      }
    });
    _this.audio = null; //当前播放器

    _this.lightThemeName = "light";
    _this.darkThemeName = "dark"; //模式配置

    _this.toggleModeName = {
      full: "full",
      mini: "mini"
    };
    _this.targetId = "music-player-controller";
    _this.openPanelPeriphery = 1; //移动差值 在 这之间 认为是点击打开panel

    _this.x = 0;
    _this.y = 0;
    _this.isDrag = false;
    var _props$playModeText = props.playModeText,
        order = _props$playModeText.order,
        orderLoop = _props$playModeText.orderLoop,
        singleLoop = _props$playModeText.singleLoop,
        shufflePlay = _props$playModeText.shufflePlay; //播放模式配置

    _this.PLAYMODE = {
      order: {
        key: "order",
        value: order
      },
      orderLoop: {
        key: "orderLoop",
        value: orderLoop
      },
      singleLoop: {
        key: "singleLoop",
        value: singleLoop
      },
      shufflePlay: {
        key: "shufflePlay",
        value: shufflePlay
      }
    };
    _this._PLAY_MODE_ = Object.values(_this.PLAYMODE);
    _this._PLAY_MODE_LENGTH_ = _this._PLAY_MODE_.length;
    return _this;
  }

  _createClass(ReactJkMusicPlayer, [{
    key: "render",
    value: function render() {
      var _this3 = this;

      var _props = this.props,
          className = _props.className,
          controllerTitle = _props.controllerTitle,
          closeText = _props.closeText,
          openText = _props.openText,
          notContentText = _props.notContentText,
          drag = _props.drag,
          style = _props.style,
          showDownload = _props.showDownload,
          showPlay = _props.showPlay,
          showReload = _props.showReload,
          showPlayMode = _props.showPlayMode,
          showThemeSwitch = _props.showThemeSwitch,
          panelTitle = _props.panelTitle,
          checkedText = _props.checkedText,
          unCheckedText = _props.unCheckedText,
          toggleMode = _props.toggleMode,
          showMiniModeCover = _props.showMiniModeCover,
          extendsContent = _props.extendsContent,
          defaultPlayMode = _props.defaultPlayMode,
          seeked = _props.seeked,
          showProgressLoadBar = _props.showProgressLoadBar,
          bounds = _props.bounds,
          defaultPosition = _props.defaultPosition,
          showMiniProcessBar = _props.showMiniProcessBar,
          preload = _props.preload,
          glassBg = _props.glassBg,
          remove = _props.remove;
      var _state = this.state,
          toggle = _state.toggle,
          playing = _state.playing,
          duration = _state.duration,
          currentTime = _state.currentTime,
          isMute = _state.isMute,
          soundValue = _state.soundValue,
          moveX = _state.moveX,
          moveY = _state.moveY,
          loading = _state.loading,
          audioListsPanelVisible = _state.audioListsPanelVisible,
          pause = _state.pause,
          theme = _state.theme,
          name = _state.name,
          cover = _state.cover,
          singer = _state.singer,
          musicSrc = _state.musicSrc,
          playId = _state.playId,
          isMobile = _state.isMobile,
          playMode = _state.playMode,
          playModeTipVisible = _state.playModeTipVisible,
          playModelNameVisible = _state.playModelNameVisible,
          initAnimate = _state.initAnimate,
          loadProgress = _state.loadProgress,
          audioLists = _state.audioLists,
          removeId = _state.removeId;
      var preloadState = preload === false || preload === "none" ? {} : preload === true ? {
        preload: "auto"
      } : {
        preload: preload
      };
      var panelToggleAnimate = initAnimate ? {
        show: audioListsPanelVisible,
        hide: !audioListsPanelVisible
      } : {
        show: audioListsPanelVisible
      };
      var _playMode_ = this.PLAYMODE[playMode || defaultPlayMode];
      var currentPlayMode = _playMode_["key"];
      var currentPlayModeName = _playMode_["value"];
      var isShowMiniModeCover = showMiniModeCover ? {
        style: {
          backgroundImage: "url(".concat(cover, ")")
        }
      } : {};

      var _currentTime = (0, _utils.formatTime)(currentTime);

      var _duration = (0, _utils.formatTime)(duration);

      var progressHandler = seeked ? {
        onChange: this.onHandleProgress,
        onAfterChange: this.onAudioSeeked
      } : {}; //进度条

      var ProgressBar = _react.default.createElement(_Slider.default, _extends({
        max: Math.ceil(duration),
        defaultValue: 0,
        value: currentTime
      }, progressHandler, sliderBaseOptions)); //下载按钮


      var DownloadComponent = showDownload ? _react.default.createElement("span", _extends({
        className: "group audio-download"
      }, _defineProperty({}, ISMOBILE ? "onTouchStart" : "onClick", this.onAudioDownload)), _react.default.createElement(_cloudDownload.default, null)) : undefined; //主题开关

      var ThemeSwitchComponent = showThemeSwitch ? _react.default.createElement("span", {
        className: "group theme-switch"
      }, _react.default.createElement(_rcSwitch.default, {
        className: "theme-switch-container",
        onChange: this.themeChange,
        checkedChildren: checkedText,
        unCheckedChildren: unCheckedText,
        checked: theme === this.lightThemeName
      })) : undefined; //重放

      var ReloadComponent = showReload ? _react.default.createElement("span", _extends({
        className: "group reload-btn"
      }, ISMOBILE ? {
        onTouchStart: this.onAudioReload
      } : {
        onClick: this.onAudioReload
      }, {
        key: "reload-btn",
        title: "Reload"
      }), _react.default.createElement(_refresh.default, null)) : undefined; //播放模式

      var PlayModeComponent = showPlayMode ? _react.default.createElement("span", _extends({
        className: (0, _classnames.default)("group loop-btn")
      }, ISMOBILE ? {
        onTouchStart: this.togglePlayMode
      } : {
        onClick: this.togglePlayMode
      }, {
        key: "play-mode-btn",
        title: this.PLAYMODE[currentPlayMode]["value"]
      }), this.renderPlayModeIcon(currentPlayMode)) : undefined;
      var miniProcessBarR = isMobile ? 30 : 40;

      var AudioController = _react.default.createElement("div", {
        className: (0, _classnames.default)("react-jinke-music-player"),
        key: "react-jinke-music-player",
        style: defaultPosition
      }, _react.default.createElement("div", {
        className: (0, _classnames.default)("music-player"),
        key: "music-player"
      }, showMiniProcessBar ? _react.default.createElement(CircleProcessBar, {
        progress: currentTime / duration,
        r: miniProcessBarR
      }) : undefined, _react.default.createElement("div", _extends({
        key: "controller",
        id: this.targetId,
        className: "scale music-player-controller"
      }, isShowMiniModeCover), loading ? _react.default.createElement(Load, null) : _react.default.createElement(_react.Fragment, null, _react.default.createElement("span", {
        className: "controller-title",
        key: "controller-title"
      }, controllerTitle), _react.default.createElement("div", {
        key: "setting",
        className: "music-player-controller-setting"
      }, toggle ? closeText : openText)))));

      return _react.default.createElement("div", {
        className: (0, _classnames.default)("react-jinke-music-player-main", {
          "light-theme": theme === this.lightThemeName,
          "dark-theme": theme === this.darkThemeName
        }, className),
        style: style
      }, toggle && isMobile ? _react.default.createElement(_PlayerMobile.default, {
        playing: playing,
        loading: loading,
        pause: pause,
        name: name,
        singer: singer,
        cover: cover,
        themeSwitch: ThemeSwitchComponent,
        duration: _duration,
        currentTime: _currentTime,
        progressBar: ProgressBar,
        onPlay: this.onPlay,
        currentPlayModeName: this.PLAYMODE[currentPlayMode]["value"],
        playMode: PlayModeComponent,
        audioNextPlay: this.audioNextPlay,
        audioPrevPlay: this.audioPrevPlay,
        playListsIcon: _react.default.createElement(_queueMusic.default, null),
        reloadIcon: ReloadComponent,
        downloadIcon: DownloadComponent,
        nextAudioIcon: _react.default.createElement(_skipNext.default, null),
        prevAudioIcon: _react.default.createElement(_skipPrevious.default, null),
        playIcon: _react.default.createElement(AnimatePlayIcon, null),
        pauseIcon: _react.default.createElement(AnimatePauseIcon, null),
        closeIcon: _react.default.createElement(_close.default, null),
        loadingIcon: _react.default.createElement(Load, null),
        playModeTipVisible: playModeTipVisible,
        openAudioListsPanel: this.openAudioListsPanel,
        onClose: this.onHidePanel,
        extendsContent: extendsContent,
        glassBg: glassBg
      }) : undefined, toggle ? undefined : drag ? _react.default.createElement(_reactDraggable.default, {
        bounds: bounds,
        position: {
          x: moveX,
          y: moveY
        },
        onDrag: this.controllerMouseMove,
        onStop: this.controllerMouseUp,
        onStart: this.controllerMouseMove
      }, AudioController) : _react.default.createElement(_react.Fragment, null, AudioController), toggle ? isMobile ? undefined : _react.default.createElement("div", {
        key: "panel",
        className: (0, _classnames.default)("music-player-panel", "translate", {
          "glass-bg": glassBg
        })
      }, _react.default.createElement("section", {
        className: "panel-content",
        key: "panel-content"
      }, _react.default.createElement("div", {
        className: (0, _classnames.default)("img-content", "img-rotate", {
          "img-rotate-pause": pause || !playing || !cover
        }),
        style: {
          backgroundImage: "url(".concat(cover, ")")
        },
        key: "img-content"
      }), _react.default.createElement("div", {
        className: "progress-bar-content",
        key: "progress-bar-content"
      }, _react.default.createElement("span", {
        className: "audio-title"
      }, name, " ", singer ? "- ".concat(singer) : ""), _react.default.createElement("section", {
        className: "audio-main"
      }, _react.default.createElement("span", {
        key: "current-time",
        className: "current-time"
      }, loading ? "--" : _currentTime), _react.default.createElement("div", {
        className: "progress-bar",
        key: "progress-bar"
      }, showProgressLoadBar ? _react.default.createElement("div", {
        className: "progress-load-bar",
        key: "progress-load-bar",
        style: {
          width: "".concat(Math.min(loadProgress, 100), "%")
        }
      }) : undefined, ProgressBar), _react.default.createElement("span", {
        key: "duration",
        className: "duration"
      }, loading ? "--" : _duration))), _react.default.createElement("div", {
        className: "player-content",
        key: "player-content"
      }, loading ? _react.default.createElement("span", null, _react.default.createElement(Load, null)) : showPlay ? _react.default.createElement("span", {
        className: "group"
      }, _react.default.createElement("span", _extends({
        className: "group prev-audio",
        title: "Previous track"
      }, ISMOBILE ? {
        onTouchStart: this.audioPrevPlay
      } : {
        onClick: this.audioPrevPlay
      }), _react.default.createElement(_skipPrevious.default, null)), _react.default.createElement("span", _extends({
        className: "group play-btn",
        key: "play-btn",
        ref: function ref(node) {
          return _this3.playBtn = node;
        }
      }, ISMOBILE ? {
        onTouchStart: this.onPlay
      } : {
        onClick: this.onPlay
      }, {
        title: playing ? "Click to pause" : "Click to play"
      }), playing ? _react.default.createElement("span", null, _react.default.createElement(AnimatePauseIcon, null)) : _react.default.createElement("span", null, _react.default.createElement(AnimatePlayIcon, null))), _react.default.createElement("span", _extends({
        className: "group next-audio",
        title: "Next track"
      }, ISMOBILE ? {
        onTouchStart: this.audioNextPlay
      } : {
        onClick: this.audioNextPlay
      }), _react.default.createElement(_skipNext.default, null))) : undefined, ReloadComponent, DownloadComponent, ThemeSwitchComponent, extendsContent && extendsContent.length >= 1 ? extendsContent.map(function (content, i) {
        return content;
      }) : undefined, _react.default.createElement("span", {
        className: "group play-sounds",
        key: "play-sound",
        title: "Volume"
      }, isMute ? _react.default.createElement("span", _extends({
        className: "sounds-icon"
      }, ISMOBILE ? {
        onTouchStart: this.onSound
      } : {
        onClick: this.onSound
      }), _react.default.createElement(_volumeOff.default, null)) : _react.default.createElement("span", _extends({
        className: "sounds-icon"
      }, ISMOBILE ? {
        onTouchStart: this.onMute
      } : {
        onClick: this.onMute
      }), _react.default.createElement(_volumeUp.default, null)), _react.default.createElement(_Slider.default, _extends({
        max: 1,
        value: soundValue,
        onChange: this.audioSoundChange,
        className: "sound-operation"
      }, sliderBaseOptions))), PlayModeComponent, _react.default.createElement("span", _extends({
        className: "group audio-lists-btn",
        key: "audio-lists-btn",
        title: "play lists"
      }, ISMOBILE ? {
        onTouchStart: this.openAudioListsPanel
      } : {
        onClick: this.openAudioListsPanel
      }), _react.default.createElement("span", {
        className: "audio-lists-icon"
      }, _react.default.createElement(_queueMusic.default, null)), _react.default.createElement("span", {
        className: "audio-lists-num"
      }, audioLists.length)), toggleMode ? _react.default.createElement("span", _extends({
        className: "group hide-panel",
        key: "hide-panel-btn"
      }, ISMOBILE ? {
        onTouchStart: this.onHidePanel
      } : {
        onClick: this.onHidePanel
      }), _react.default.createElement(_minusSquareO.default, null)) : undefined)), _react.default.createElement(PlayModel, {
        visible: playModelNameVisible,
        value: currentPlayModeName
      })) : undefined, _react.default.createElement(_AudioListsPanel.default, {
        playId: playId,
        pause: pause,
        loading: loading ? _react.default.createElement(Load, null) : undefined,
        visible: audioListsPanelVisible,
        audioLists: audioLists,
        notContentText: notContentText,
        onPlay: this.audioListsPlay,
        onCancel: this.closeAudioListsPanel,
        playIcon: _react.default.createElement(AnimatePlayIcon, null),
        pauseIcon: _react.default.createElement(AnimatePauseIcon, null),
        closeIcon: _react.default.createElement(_close.default, null),
        panelTitle: panelTitle,
        isMobile: ISMOBILE,
        panelToggleAnimate: panelToggleAnimate,
        glassBg: glassBg,
        cover: cover,
        remove: remove,
        deleteIcon: _react.default.createElement(_trashO.default, null),
        onDelete: this.deleteAudioLists,
        removeId: removeId,
        audioListsDragEnd: this.audioListsDragEnd
      }), _react.default.createElement("audio", _extends({
        key: "audio",
        className: "music-player-audio"
      }, preloadState, {
        src: musicSrc,
        ref: function ref(node) {
          return _this3.audio = node;
        }
      })));
    } //播放模式切换

  }, {
    key: "getBaseAudioInfo",
    //返回给使用者的 音乐信息
    value: function getBaseAudioInfo() {
      var _state2 = this.state,
          cover = _state2.cover,
          name = _state2.name,
          musicSrc = _state2.musicSrc,
          soundValue = _state2.soundValue;
      var _audio = this.audio,
          currentTime = _audio.currentTime,
          duration = _audio.duration,
          muted = _audio.muted,
          networkState = _audio.networkState,
          readyState = _audio.readyState,
          played = _audio.played,
          paused = _audio.paused,
          ended = _audio.ended,
          startDate = _audio.startDate;
      return {
        cover: cover,
        name: name,
        musicSrc: musicSrc,
        volume: soundValue,
        currentTime: currentTime,
        duration: duration,
        muted: muted,
        networkState: networkState,
        readyState: readyState,
        played: played,
        paused: paused,
        ended: ended,
        startDate: startDate
      };
    } //播放

  }, {
    key: "componentWillReceiveProps",
    //当父组件 更新 props 时 如 audioLists 改变 更新播放信息
    value: function componentWillReceiveProps(_ref11) {
      var audioLists = _ref11.audioLists;

      if (!(0, _utils.arrayEqual)(audioLists)(this.props.audioLists)) {
        //当列表改变 如3首变成1首 这时以最小值播放
        var playId = Math.min(audioLists.length, this.state.playId);
        this.initPlayInfo(audioLists, playId);
        this.bindEvents(this.audio);
        this.props.onAudioListsChange && this.props.onAudioListsChange(playId, audioLists, this.getBaseAudioInfo());
      }
    } //合并state 更新初始值

  }, {
    key: "componentWillMount",
    value: function componentWillMount() {
      var _props2 = this.props,
          theme = _props2.theme,
          mode = _props2.mode,
          audioLists = _props2.audioLists,
          defaultPlayMode = _props2.defaultPlayMode,
          remember = _props2.remember,
          defaultPlayIndex = _props2.defaultPlayIndex; //切换 'mini' 或者 'full' 模式

      this.toggleMode(mode);

      if (audioLists.length >= 1) {
        var playId = this.getDefaultPlayIndex(defaultPlayIndex);
        var lastPlayStatus = remember ? this.getLastPlayStatus() : {
          playId: playId,
          playMode: defaultPlayMode
        };
        this.setState(_objectSpread({}, this.getPlayInfo(audioLists, playId), lastPlayStatus));
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this.unBindEvents(this.audio, undefined, false);
      this.media.removeListener(this.listenerIsMobile);
      this.media = undefined;
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      this.addMobileListener();
      this.setDefaultAudioVolume();

      if (this.props.audioLists.length >= 1) {
        this.bindEvents(this.audio);

        if (ISMOBILE) {
          this.bindMobileAutoPlayerEvents();
        }
      }
    }
  }]);

  return ReactJkMusicPlayer;
}(_react.PureComponent);

exports.default = ReactJkMusicPlayer;
Object.defineProperty(ReactJkMusicPlayer, "defaultProps", {
  configurable: true,
  enumerable: true,
  writable: true,
  value: {
    audioLists: [],
    theme: "dark",
    mode: "mini",
    playModeText: {
      order: "order",
      orderLoop: "orderLoop",
      singleLoop: "singleLoop",
      shufflePlay: "shufflePlay"
    },
    defaultPlayMode: "order",
    defaultPosition: {
      left: 0,
      top: 0
    },
    controllerTitle: _react.default.createElement(_headphones.default, null),
    panelTitle: "PlayList",
    closeText: "close",
    openText: "open",
    notContentText: "no music",
    checkedText: "",
    unCheckedText: "",
    once: false,
    //onAudioPlay 事件  是否只触发一次
    drag: true,
    toggleMode: true,
    //能换在迷你 和完整模式下 互相切换
    showMiniModeCover: true,
    //迷你模式下 是否显示封面图
    showDownload: true,
    showPlay: true,
    showReload: true,
    showPlayMode: true,
    showThemeSwitch: true,
    playModeTipVisible: false,
    //手机端切换播放模式
    autoPlay: true,
    defaultVolume: 100,
    showProgressLoadBar: true,
    //音频预加载进度
    seeked: true,
    playModeShowTime: 600,
    //播放模式提示 显示时间,
    bounds: "body",
    //移动边界
    showMiniProcessBar: false,
    //是否在迷你模式 显示进度条
    loadAudioErrorPlayNext: true,
    // 加载音频失败时 是否尝试播放下一首
    preload: false,
    //是否在页面加载后立即加载音频
    glassBg: false,
    //是否是毛玻璃效果
    remember: false,
    //是否记住当前播放状态
    remove: true,
    //音乐是否可以删除
    defaultPlayIndex: 0 //默认播放索引

  }
});
Object.defineProperty(ReactJkMusicPlayer, "propTypes", {
  configurable: true,
  enumerable: true,
  writable: true,
  value: {
    audioLists: _propTypes.default.array.isRequired,
    theme: _propTypes.default.oneOf(["dark", "light"]),
    mode: _propTypes.default.oneOf(["mini", "full"]),
    defaultPlayMode: _propTypes.default.oneOf(["order", "orderLoop", "singleLoop", "shufflePlay"]),
    drag: _propTypes.default.bool,
    seeked: _propTypes.default.bool,
    autoPlay: _propTypes.default.bool,
    playModeText: _propTypes.default.object,
    panelTitle: _propTypes.default.string,
    closeText: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.object]),
    openText: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.object]),
    notContentText: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.object]),
    controllerTitle: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.object]),
    defaultPosition: _propTypes.default.shape({
      top: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
      left: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
      right: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
      bottom: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string])
    }),
    onAudioPlay: _propTypes.default.func,
    onAudioPause: _propTypes.default.func,
    onAudioEnded: _propTypes.default.func,
    onAudioAbort: _propTypes.default.func,
    onAudioVolumeChange: _propTypes.default.func,
    onAudioLoadError: _propTypes.default.func,
    onAudioProgress: _propTypes.default.func,
    onAudioSeeked: _propTypes.default.func,
    onAudioDownload: _propTypes.default.func,
    onAudioReload: _propTypes.default.func,
    onThemeChange: _propTypes.default.func,
    onAudioListsChange: _propTypes.default.func,
    onPlayModeChange: _propTypes.default.func,
    onModeChange: _propTypes.default.func,
    onAudioListsPanelChange: _propTypes.default.func,
    onAudioPlayTrackChange: _propTypes.default.func,
    onAudioListsDragEnd: _propTypes.default.func,
    showDownload: _propTypes.default.bool,
    showPlay: _propTypes.default.bool,
    showReload: _propTypes.default.bool,
    showPlayMode: _propTypes.default.bool,
    showThemeSwitch: _propTypes.default.bool,
    showMiniModeCover: _propTypes.default.bool,
    toggleMode: _propTypes.default.bool,
    once: _propTypes.default.bool,
    extendsContent: _propTypes.default.array,
    checkedText: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.object]),
    unCheckedText: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.object]),
    defaultVolume: _propTypes.default.number,
    playModeShowTime: _propTypes.default.number,
    bounds: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.object]),
    showMiniProcessBar: _propTypes.default.bool,
    loadAudioErrorPlayNext: _propTypes.default.bool,
    preload: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.oneOf(["auto", "metadata", "none"])]),
    glassBg: _propTypes.default.bool,
    remember: _propTypes.default.bool,
    remove: _propTypes.default.bool,
    defaultPlayIndex: _propTypes.default.number
  }
});
